pkg/ddevapp/craftcms.go

```diff
@@ -27,7 +27,9 @@ func craftCmsConfigOverrideAction(app *DdevApp) error {
 
 // Returns the upload directory for importing files, if not already set
 func getCraftCmsUploadDir(app *DdevApp) string {
-	app.UploadDir = "files"
+	if app.UploadDir == "" {
+		return "files"
```

Comment:
I don't think defaulting to "files" is a reasonable thing to do, given our conversation. It shouldn't have a default if there's no such expectation in Craft.

Reply:
Great, how do we get that to work with the testing framework?

Reply:
I think there should be a default. The issue this PR is intending to fix is to allow the `upload_dir` to be overridden via config, which is a perfectly reasonable thing to want.

Providing a default in this context is fine; we just need to ensure that there is a way the user can _override_ that default.

The issue I had was the `if app.UploadDir == "" {` always failed, which means it is set to _something_ by default (or maybe it's null?). If we can figure out what that default value is, we can either:

* Properly check for it, and not set a default ourselves
* Use that default value in our tests

The latter of the two requires redoing the test project setup, so I'd have a preference for the first

ref: https://github.com/drud/ddev/issues/4302

Reply:
Trying to look at it this morning. We may want to add the upload_dir to the testing framework.

Reply:
@khalwat I've asked about this a number of times and the answer was always "there is no default files dir in Craft". So there are two things at play right? We obviously have to make upload_dir work right and override any project-type-provided upload_dir. And we also have to do some reasonable testing.

Reply:
Correct, there is no default for Craft, because people can organize their project any way they want.

Also what you might typically think of as an "upload_dir" might be somewhat of a dated or at least very CMS-specific concept, but generally relates to where user uploaded files (images, etc.) will go, correct?

In Craft, the default project has zero Asset Volumes set up. You set them up yourself, and you choose where they point to, whether local directories, or S3 buckets, or other cloud storage.

So we are providing a default of `files/` for two reasons:

1) We need _something_ for the test project to pass the "uploading file" test, which still seems useful to me
2) It's a sane default that at least some % of people using Craft will use.

Thus I think it's fine to provide said default _providing_ that we can properly allow them to override the default via the `.yaml` config.

The primary issue I was having, and @bencroker has run into, is that we can't do this test:

```go
if app.UploadDir == ""
```

...because it always fails. A default DDEV project has `app.UploadDir` set to _something_ (or maybe null?) which causes that check to fail, which then causes us to not be able to default it to `files/` which then causes the test project to fail, because it's expecting a file to be in `upload_dir` (which in the project is `files/`)

The only other option I can think of is some way to default _just the test project_ to use an `upload_dir` of `files/` but I'm not sure how to do that, and I'm sure involves some redoing of the test project?

Reply:
All that said, if you want to instead change the testing framework, that's fine too. I'm just thinking that could require a good bit more work, in that we'd have to ensure that it doesn't break any of the other test projects, or refactor the other test projects as well.